spoc/snipping/snipping_strategies.py

```diff
@@ -0,0 +1,240 @@
+            SUM(p.contact_count) as contacts
+        FROM {source_table} as p
+        INNER JOIN chunk as t ON t.chrom = p.chrom
+            and
+                abs(FLOOR((p.start_1 - t.pos)/{bin_size}::float))::int <= {pixel_offset}
+            and
+                abs(FLOOR((p.start_2 - t.pos)/{bin_size}::float))::int <= {pixel_offset}
+            and
+                abs(FLOOR((p.start_3 - (t.pos + {relative_offset}))/{bin_size}::float))::int <= {cis_selector_offset}
+        GROUP BY 1,2,3
+    """
+
+    def __init__(
+        self,
+        bin_size: int,
+        half_window_size: int,
+        snipping_value: SnippingValues,
+        **kwargs,
+    ):
+        """Override constructor to add additional parameters."""
+        super().__init__(bin_size, half_window_size, snipping_value, **kwargs)
+        self._position_slack: int = kwargs.get(
+            "position_slack", self._bin_size
+        )  # default is binsize
+        self._relative_offset: int = kwargs.get("relative_offset", 0)
+        self._expected = None
+
+    def get_params(self) -> dict:
+        return {
+            "bin_size": self._bin_size,
+            "half_window_size": self._half_window_size,
+            "snipping_value": self._snipping_value,
+            "position_slack": self._position_slack,
+            "relative_offset": self._relative_offset,
+        }
+
+    def _align_positions_to_bins(self, trans_positions: pd.DataFrame):
+        """Adds index and round positions to bins"""
+        if "pos" not in trans_positions.columns:
+            trans_positions = trans_positions.assign(
+                pos=lambda df_: (df_.start + df_.end) // 2
+            )
+        return trans_positions.assign(
+            position_id=lambda df_: range(len(df_)),
+            pos=lambda df_: (df_.pos // self._bin_size) * self._bin_size,
+        )
+
+    def _get_array_coordinates_from_offset(self, offset: pd.Series):
+        """Transform offsets to start from 0 to be used as array index"""
+        return (offset + (self._half_window_size // self._bin_size)).astype(int).values
+
+    def _reduce_snipping_frame(self, snips: pd.DataFrame):
+        """Takes concat result of snipping and reduces
+        it along the region dimension"""
+        output_size = 2 * (self._half_window_size // self._bin_size) + 1
+        return (
+            COO(
+                (
+                    snips.position_id.values,
+                    self._get_array_coordinates_from_offset(snips.offset_1),
+                    self._get_array_coordinates_from_offset(snips.offset_2),
+                ),
+                snips.contacts.values,
+                # TODO: fix shape such that if ids are missing from the end of the input, input shape will not be affected
+                shape=(np.max(snips.position_id) + 1, output_size, output_size),
+            )
+            .mean(axis=0) # this reduces the result along the region id dimension
```

Comment:
If snips for some positions are totally missing, then this will produce an incorrect mean. It makes sense to replace it with `.sum(axis=0` and then divide by the precalculated group size.